File: frontend/src/api/portfolio.ts
import type { PortfolioItem, PortfolioSourceType } from '../types'
import { apiFetch } from './client'

export interface BackendPortfolioItem {
  id: number
  profile_id: number
  title: string
  url: string
  source_type: string
  main_category: string
  sub_category: string
  industry_tags: string[]
  skill_tags: string[]
  style_tags: string[]
  tools_tags: string[]
  description: string | null
  priority_score: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface PortfolioTaxonomy {
  taxonomy: Record<string, string[]>
  source_types: string[]
}

export interface PortfolioFilters {
  mainCategory?: string
  subCategory?: string
  sourceType?: string
  tag?: string
  search?: string
}

export function mapPortfolioItem(raw: BackendPortfolioItem): PortfolioItem {
  return {
    id: raw.id,
    profileId: raw.profile_id,
    title: raw.title,
    url: raw.url,
    sourceType: raw.source_type as PortfolioSourceType,
    mainCategory: raw.main_category,
    subCategory: raw.sub_category,
    industryTags: raw.industry_tags ?? [],
    skillTags: raw.skill_tags ?? [],
    styleTags: raw.style_tags ?? [],
    toolsTags: raw.tools_tags ?? [],
    description: raw.description,
    priorityScore: raw.priority_score,
    isActive: raw.is_active,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  }
}

export async function getPortfolioTaxonomy(): Promise<PortfolioTaxonomy> {
  return apiFetch<PortfolioTaxonomy>('/portfolio/taxonomy')
}

export async function listPortfolioItems(
  profileId: number,
  filters: PortfolioFilters = {},
): Promise<PortfolioItem[]> {
  const params = new URLSearchParams()
  if (filters.mainCategory) params.set('main_category', filters.mainCategory)
  if (filters.subCategory) params.set('sub_category', filters.subCategory)
  if (filters.sourceType) params.set('source_type', filters.sourceType)
  if (filters.tag) params.set('tag', filters.tag)
  if (filters.search) params.set('search', filters.search)

  const qs = params.toString()
  const path = `/profiles/${profileId}/portfolio${qs ? `?${qs}` : ''}`
  const rows = await apiFetch<BackendPortfolioItem[]>(path)
  return rows.map(mapPortfolioItem)
}

export async function getPortfolioItem(id: number): Promise<PortfolioItem> {
  const row = await apiFetch<BackendPortfolioItem>(`/portfolio/${id}`)
  return mapPortfolioItem(row)
}

export interface PortfolioItemPayload {
  title: string
  url: string
  sourceType: PortfolioSourceType
  mainCategory: string
  subCategory: string
  industryTags?: string[]
  skillTags?: string[]
  styleTags?: string[]
  toolsTags?: string[]
  description?: string
  priorityScore?: number
}

function toBackendPayload(payload: PortfolioItemPayload) {
  return {
    title: payload.title,
    url: payload.url,
    source_type: payload.sourceType,
    main_category: payload.mainCategory,
    sub_category: payload.subCategory,
    industry_tags: payload.industryTags ?? [],
    skill_tags: payload.skillTags ?? [],
    style_tags: payload.styleTags ?? [],
    tools_tags: payload.toolsTags ?? [],
    description: payload.description ?? null,
    priority_score: payload.priorityScore ?? 0,
  }
}

export async function createPortfolioItem(
  profileId: number,
  payload: PortfolioItemPayload,
): Promise<PortfolioItem> {
  const row = await apiFetch<BackendPortfolioItem>(`/profiles/${profileId}/portfolio`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(toBackendPayload(payload)),
  })
  return mapPortfolioItem(row)
}

export async function updatePortfolioItem(
  id: number,
  payload: Partial<PortfolioItemPayload> & { isActive?: boolean },
): Promise<PortfolioItem> {
  const row = await apiFetch<BackendPortfolioItem>(`/portfolio/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      title: payload.title,
      url: payload.url,
      source_type: payload.sourceType,
      main_category: payload.mainCategory,
      sub_category: payload.subCategory,
      industry_tags: payload.industryTags,
      skill_tags: payload.skillTags,
      style_tags: payload.styleTags,
      tools_tags: payload.toolsTags,
      description: payload.description,
      priority_score: payload.priorityScore,
      is_active: payload.isActive,
    }),
  })
  return mapPortfolioItem(row)
}

export async function deletePortfolioItem(id: number): Promise<void> {
  await apiFetch(`/portfolio/${id}`, { method: 'DELETE' })
}

export interface PortfolioAnalyzeResult {
  main_category: string
  sub_category: string
  industry_tags: string[]
  skill_tags: string[]
  style_tags: string[]
  tools_tags: string[]
  best_for_jobs: string[]
  short_summary: string
  confidence_score: number
}

export interface PortfolioAnalyzeInput {
  title?: string
  url?: string
  description?: string
  sourceType?: string
}

function mapAnalyzeResult(raw: PortfolioAnalyzeResult) {
  return {
    mainCategory: raw.main_category ?? '',
    subCategory: raw.sub_category ?? '',
    industryTags: raw.industry_tags ?? [],
    skillTags: raw.skill_tags ?? [],
    styleTags: raw.style_tags ?? [],
    toolsTags: raw.tools_tags ?? [],
    bestForJobs: raw.best_for_jobs ?? [],
    shortSummary: raw.short_summary ?? '',
    confidenceScore: raw.confidence_score ?? 0,
  }
}

export async function analyzePortfolioItem(
  input: PortfolioAnalyzeInput,
  itemId?: number,
) {
  const path =
    itemId != null ? `/portfolio/${itemId}/analyze` : '/portfolio/analyze'
  const raw = await apiFetch<PortfolioAnalyzeResult>(path, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      title: input.title,
      url: input.url,
      description: input.description,
      source_type: input.sourceType,
    }),
  })
  return mapAnalyzeResult(raw)
}

File: backend/app/routes/portfolio.py
import logging

from fastapi import APIRouter, Depends

from app.constants.portfolio_taxonomy import PORTFOLIO_SOURCE_TYPES, PORTFOLIO_TAXONOMY
from app.intelligence.portfolio_analyze import analyze_portfolio_item
from app.models_portfolio import PortfolioItem
from app.portfolio.utils import normalize_tags, validate_source_type, validate_taxonomy
from app.schemas_portfolio import (
    PortfolioAnalyzeInput,
    PortfolioAnalyzeResult,
    PortfolioItemRead,
    PortfolioItemUpdate,
    PortfolioTaxonomyRead,
)
from app.security.scoping import UserScope, get_scope

logger = logging.getLogger(__name__)

router = APIRouter(tags=["portfolio"])


@router.get("/portfolio/taxonomy", response_model=PortfolioTaxonomyRead)
def get_portfolio_taxonomy() -> PortfolioTaxonomyRead:
    return PortfolioTaxonomyRead(
        taxonomy=PORTFOLIO_TAXONOMY,
        source_types=list(PORTFOLIO_SOURCE_TYPES),
    )


@router.post("/portfolio/analyze", response_model=PortfolioAnalyzeResult)
def analyze_portfolio_draft(
    body: PortfolioAnalyzeInput,
    scope: UserScope = Depends(get_scope),
) -> PortfolioAnalyzeResult:
    """Suggest categories and tags for an unsaved portfolio item (does not persist)."""
    result = analyze_portfolio_item(
        scope.session,
        user_id=scope.user_id,
        title=body.title or "",
        url=body.url or "",
        description=body.description,
        source_type=body.source_type,
    )
    return PortfolioAnalyzeResult(**result)


@router.post("/portfolio/{item_id}/analyze", response_model=PortfolioAnalyzeResult)
def analyze_portfolio_by_id(
    item_id: int,
    body: PortfolioAnalyzeInput | None = None,
    scope: UserScope = Depends(get_scope),
) -> PortfolioAnalyzeResult:
    """Suggest categories and tags for a saved item; body fields override stored values."""
    item = scope.get_or_404(PortfolioItem, item_id, detail="Portfolio item not found.")

    overrides = body or PortfolioAnalyzeInput()
    result = analyze_portfolio_item(
        scope.session,
        user_id=scope.user_id,
        title=(overrides.title or item.title).strip(),
        url=(overrides.url or item.url).strip(),
        description=overrides.description if overrides.description is not None else item.description,
        source_type=overrides.source_type or item.source_type,
    )
    return PortfolioAnalyzeResult(**result)


@router.get("/portfolio/{item_id}", response_model=PortfolioItemRead)
def get_portfolio_item(item_id: int, scope: UserScope = Depends(get_scope)) -> PortfolioItemRead:
    return scope.get_or_404(PortfolioItem, item_id, detail="Portfolio item not found.")


@router.put("/portfolio/{item_id}", response_model=PortfolioItemRead)
def update_portfolio_item(
    item_id: int,
    body: PortfolioItemUpdate,
    scope: UserScope = Depends(get_scope),
) -> PortfolioItemRead:
    item = scope.get_or_404(PortfolioItem, item_id, detail="Portfolio item not found.")

    main_category = body.main_category if body.main_category is not None else item.main_category
    sub_category = body.sub_category if body.sub_category is not None else item.sub_category
    if body.main_category is not None or body.sub_category is not None:
        validate_taxonomy(main_category, sub_category)

    if body.source_type is not None:
        validate_source_type(body.source_type)

    if body.title is not None:
        item.title = body.title.strip()
    if body.url is not None:
        item.url = body.url.strip()
    if body.source_type is not None:
        item.source_type = body.source_type
    if body.main_category is not None:
        item.main_category = body.main_category
    if body.sub_category is not None:
        item.sub_category = body.sub_category
    if body.industry_tags is not None:
        item.industry_tags = normalize_tags(body.industry_tags)
    if body.skill_tags is not None:
        item.skill_tags = normalize_tags(body.skill_tags)
    if body.style_tags is not None:
        item.style_tags = normalize_tags(body.style_tags)
    if body.tools_tags is not None:
        item.tools_tags = normalize_tags(body.tools_tags)
    if body.description is not None:
        item.description = body.description.strip() or None
    if body.priority_score is not None:
        item.priority_score = body.priority_score
    if body.is_active is not None:
        item.is_active = body.is_active

    scope.session.commit()
    scope.session.refresh(item)
    return item


@router.delete("/portfolio/{item_id}")
def delete_portfolio_item(item_id: int, scope: UserScope = Depends(get_scope)) -> dict[str, object]:
    item = scope.get_or_404(PortfolioItem, item_id, detail="Portfolio item not found.")
    item.is_active = False
    scope.session.commit()
    return {"id": item_id, "is_active": False}
